tools/myst/enc/enc.c

```diff
@@ -143,37 +128,60 @@ static uint64_t _forward_exception_as_signal_to_kernel(
         // Note that the si_addr in the simulated #PF always has the
         // lower 12 bits cleared.
         siginfo.si_addr = (void*)oe_exception_record->faulting_address;
-        (*_kargs.myst_handle_host_signal)(&siginfo, &_mcontext);
-        _mcontext_to_oe_context(&_mcontext, oe_context);
-        return OE_EXCEPTION_CONTINUE_EXECUTION;
+        supported_code = true;
     }
-    if (oe_exception_code == OE_EXCEPTION_ACCESS_VIOLATION)
+    else if (oe_exception_code == OE_EXCEPTION_ACCESS_VIOLATION)
     {
         // #GP can only be delivered on icelake.
         siginfo.si_code = SEGV_ACCERR;
         siginfo.si_signo = SIGSEGV;
         // `si_addr` is always null for #GP.
         siginfo.si_addr = NULL;
-        (*_kargs.myst_handle_host_signal)(&siginfo, &_mcontext);
-        _mcontext_to_oe_context(&_mcontext, oe_context);
-        return OE_EXCEPTION_CONTINUE_EXECUTION;
+        supported_code = true;
     }
-    if (oe_exception_code == OE_EXCEPTION_X87_FLOAT_POINT)
+    else if (oe_exception_code == OE_EXCEPTION_X87_FLOAT_POINT)
     {
         // ATTN: Consider implementing accurate si-code for
         // OE_EXCEPTION_X87_FLOAT_POINT
         siginfo.si_code = FPE_FLTINV;
         siginfo.si_signo = SIGFPE;
-        (*_kargs.myst_handle_host_signal)(&siginfo, &_mcontext);
-        _mcontext_to_oe_context(&_mcontext, oe_context);
-        return OE_EXCEPTION_CONTINUE_EXECUTION;
+        supported_code = true;
     }
-    if (oe_exception_code == OE_EXCEPTION_DIVIDE_BY_ZERO)
+    else if (oe_exception_code == OE_EXCEPTION_DIVIDE_BY_ZERO)
     {
         siginfo.si_code = FPE_INTDIV;
         siginfo.si_signo = SIGFPE;
-        (*_kargs.myst_handle_host_signal)(&siginfo, &_mcontext);
-        _mcontext_to_oe_context(&_mcontext, oe_context);
+        supported_code = true;
+    }
+
+    if (supported_code)
+    {
+        /* The OE layer hardware exception handling scheme requires
+         * any registered handler call to return such that the OE
+         * can clean up internal states after the handler finishes.
+         * However, invoking a signal handler of the programming
+         * language or the application does not always meet the
+         * requirement as the handler might not return. Therefore,
+         * we set up the oe_context and return to OE layer with
+         * OE_EXCEPTION_CONTINUE_EXECUTION. After OE performs
+         * necessary cleanup, the execution will "jump to"
+         * myst_handle_host_signal using Mystikos stack.
+         */
+
+        uint64_t rsp;
+        uint64_t rbp;
+
+        asm volatile("mov %%rsp, %0" : "=r"(rsp));
+        asm volatile("mov %%rbp, %0" : "=r"(rbp));
+
+        oe_context->rip = _kargs.myst_handle_host_signal;
+        // mimic the behavior of call (i.e., substract 8 bytes) to ensure the
+        // stack alignment
+        oe_context->rsp = rsp - 8;
```

Comment:
RSP could be any value (e.g odd number) here. It is safer to align using AND operation as we do in enter.S.

Reply:
done